router/proxy/transport.go

```diff
@@ -64,13 +66,33 @@ func (t *transport) setStickyBackend(res *http.Response, originalStickyBackend s
 	}
 }
 
-// always sets the response.Request.URL.Host to the last backend that was
-// connected to.
-func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+func (t *transport) RoundTrip(ctx context.Context, req *http.Request) (*http.Response, error) {
 	// http.Transport closes the request body on a failed dial, issue #875
 	req.Body = &fakeCloseReadCloser{req.Body}
 	defer req.Body.(*fakeCloseReadCloser).RealClose()
 
+	resultc := make(chan httpResult, 1)
+
+	go func() {
+		defer close(resultc)
+		res, err := t.roundTrip(req)
+		resultc <- httpResult{res, err}
+	}()
+
+	select {
+	case <-ctx.Done():
+		httpTransport.CancelRequest(req)
```

Comment:
cool. i suspect it will be easier to follow if the `for` loop is moved into `RoundTrip` and the `resultc/go/select` for `httpTransport.RoundTrip` stuff is moved to `roundTrip`.